utils/swift_build_support/swift_build_support/products/product.py

```diff
@@ -0,0 +1,32 @@
+# swift_build_support/products/product.py -----------------------*- python -*-
+#
+# This source file is part of the Swift.org open source project
+#
+# Copyright (c) 2014 - 2016 Apple Inc. and the Swift project authors
+# Licensed under Apache License v2.0 with Runtime Library Exception
+#
+# See http://swift.org/LICENSE.txt for license information
+# See http://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+#
+# ----------------------------------------------------------------------------
+
+
+class Product(object):
+    @classmethod
+    def product_name(klass):
+        """product_name() -> str
+
+        The identifier-style name to use for this product.
+        """
+        return klass.__name__.lower()
+
+    @classmethod
+    def get_build_directory_name(klass, host_target):
+        return "{}-{}".format(klass.product_name(),
+                              host_target.name)
+
+    def __init__(self, args, toolchain, source_dir, build_dir):
+        self.args = args
```

Comment:
Can we remove the `args` part of this initialiser back to Ninja? I don't like the idea of each product parsing its own options from the build-script; it makes it easy to lose track of which options are required and for what. For and example of where this is already a problem, see `CMake` (cmake.py); it also parses its own build-script arguments, and for reasons I can't explain is setting sanitizers and compiler version strings (both of which only make sense for LLVM/Swift products). I feel that something like that would be easier to spot if `CMake()` took a `clang_compiler_version` argument. Somebody working on the general build-script would realise that it doesn't make sense for the general CMake object to require that information.

Basically, I think we should avoid splitting the argument parsing logic up from one file (`build-script`), to a collection of many files.